### PR TITLE
Dynamic strong params for nested attributes

### PR DESCRIPTION
**Pull request by bear-in-mind**
Following our wish to ease user's life as much as possible, and `strong_params` still being a real pain to write and configure for nested attributes, here's a first step towards that goal.

## In the model
Capture d’écran 2021-02-07 à 22 37 07
For each abymized resource, we can now either : 

- Do nothing, like before. Nested attributes need to be manually added in the strong params
- Pass an array of attributes to allow
- Pass the `all_attributes` option to automatically allow all attributes (except: `created_at`, `updated_at`)
- Pass an array of attributes to reject

Capture d’écran 2021-02-19 à 18 05 46

These then get stored in an instance variable of the `Abyme::Model` module, which can be accessed for each class through the `Klass.abyme_attributes` method, which in turn renders a hash of attributes, formatted exactly the way strong_params require them.

## In the controller

Thanks to the `Abyme::Controller` module, we now have access to something like this in the controller : 
Capture d’écran 2021-02-20 à 15 26 11

This method tries to parse the ressource on which the current controller is based ; if it exists, it will automatically call the `abyme_attributes` on the relevant model class.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,12 +1,29 @@
...
+# views/projects/_form.html.erb
+<%= simple_form_for @project do |f| %>
+  <%= f.input :title %>
+  <%= f.input :description %>
+  <%= f.submit 'Save' %>
+
+  <%= f.abyme_for(:tasks) %>
+<% end %>
+```
+Supposing you have a partial located in `views/abyme/_task_fields` containing your fields for `tasks`, this command will generate and display 3 elements in this order :
```
Comment (charlottepages) [inline README.md:19]: unclear which command you are referring to
```suggestion
Supposing you have a partial located in `views/abyme/_task_fields` containing your fields for `tasks`, the `abyme_for` command will generate and display 3 elements in this order :
```

File: README.md [hunk 2]
```diff
@@ -142,15 +159,15 @@ abyme provides helper methods for all these. Here's how our form for `Project` l
...
+  <%= f.abyme_for(:tasks) do |abyme| %>
     <%= abyme.records %>
     <%= abyme.new_records %>
     <%= add_association %>
   <% end %>
 <% end %>
 ```
 
-`abyme.records` will contain the persisted associations fields, while `abyme.new_records` will contain fields for the new associations. `add_association` will by default generate a button with a text of type "Add `resource_name`". To work properly, this method **has** to be called **inside the block** passed to the `abymize` method.
+`abyme.records` will contain the persisted associations fields, while `abyme.new_records` will contain fields for the new associations. `add_association` will by default generate a button with a text of type "Add `resource_name`". To work properly, this method **has** to be called **inside the block** passed to the `abyme_for` method.
```
Comment (charlottepages) [inline README.md:170]: sorry wording again ^^ : unsure about `associations` 
which imo refers to the connection between two models 
wb: `associated records` or `child records` ?
👉 idem for `add_association`

File: README.md [hunk 3]
```diff
@@ -202,31 +219,31 @@ Which is the way you would configure `nested_attributes` 90% of the time. Should
...
 ```ruby
-  <%= abymize(:tasks, f) do |abyme| %>
+  <%= f.abyme_for(:tasks) do |abyme| %>
     <%= abyme.records(order: { created_at: :asc }) %>
     <%= abyme.new_records %>
     <%= add_association %>
   <% end %>
 ```
-* `partial:` : allows you to indicate a custom partial, if one has not already been passed to `abymize`.
+* `partial:` : allows you to indicate a custom partial, if one has not already been passed to `abyme_for`.
```
Comment (charlottepages) [inline README.md:236]: can you also write this then: 
```
<%= f.abyme_for(:tasks, partial: 'projects/task_fields') do |abyme| %>
  <%= abyme.records(order: { created_at: :asc }) %>
  <%= abyme.new_records %>
  <%= add_association %>
<% end %>
```
?

File: spec/dummy/app/controllers/projects_controller.rb [hunk 4]
```diff
@@ -25,21 +25,28 @@ def update
...
-        :_destroy, 
-        comments_attributes: [
-          :id, 
-          :content, 
-          :_destroy
-        ]
-      ],
-      participants_attributes: [:id, :email, :_destroy]
+    params.require(:project).permit(
+      :title, :description, Project.abyme_attributes
```
Comment (charlottepages) [inline spec/dummy/app/controllers/projects_controller.rb:49]: could be cool to have something like:
```suggestion
      :title, :description, abyme_attributes
```

**Review by charlottepages — COMMENTED** [review]
comments only as I understand we'll need to implement all 'TODO' before merging anyway ;)

File: spec/dummy/app/controllers/projects_controller.rb (hunk 4 above)
Comment (bear-in-mind) [reply spec/dummy/app/controllers/projects_controller.rb:49]: C'est un one shot !

**Review by ghbozz — APPROVED** [review]

**Pull request merged**